src/dsa/rpo_falcon512/signature.rs

```diff
@@ -39,251 +42,321 @@ use super::{
+        let c = hash_to_point_rpo256(message, &self.nonce);
 
-        // Make sure the signature is valid
-        let s2 = self.sig_poly();
-        let c = self.hash_to_point(message);
+        let s1_fft = self.s1.fft();
+        let s2_fft = self.s2.fft();
+        let c_fft = c.fft();
 
-        let s1 = c - s2 * h;
+        // recover the public key polynomial using h = s2^(-1) * (c - s1)
+        let h_fft = (c_fft - s1_fft).hadamard_div(&s2_fft);
+        let h = h_fft.ifft();
 
-        let sq_norm = s1.sq_norm() + s2.sq_norm();
-        sq_norm <= SIG_L2_BOUND
+        // compute the hash of the public key polynomial
+        let h_felt: Polynomial<Felt> = h.clone().into();
+        let h_digest: Word = Rpo256::hash_elements(&h_felt.coefficients).into();
+
+        h_digest == pubkey_com && verify_helper(&c, &self.s2, &h.into())
     }
 }
 
-// SERIALIZATION / DESERIALIZATION
-// ================================================================================================
-
 impl Serializable for Signature {
     fn write_into<W: ByteWriter>(&self, target: &mut W) {
-        target.write_bytes(&self.pk);
-        target.write_bytes(&self.sig);
+        target.write(&self.header);
+        target.write(&self.nonce);
+        target.write(&self.s1);
+        target.write(&self.s2);
     }
 }
 
 impl Deserializable for Signature {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
-        let pk: PublicKeyBytes = source.read_array()?;
-        let sig: SignatureBytes = source.read_array()?;
-
-        // make sure public key and signature can be decoded correctly
-        let pk_polynomial = Polynomial::from_pub_key(&pk)
-            .map_err(|err| DeserializationError::InvalidValue(err.to_string()))?
-            .into();
-        let sig_polynomial = Polynomial::from_signature(&sig)
-            .map_err(|err| DeserializationError::InvalidValue(err.to_string()))?
-            .into();
-
-        Ok(Self { pk, sig, pk_polynomial, sig_polynomial })
+        let header = source.read()?;
+        let nonce = source.read()?;
+        let s1 = source.read()?;
+        let s2 = source.read()?;
+
+        Ok(Self { header, nonce, s1, s2 })
     }
 }
 
-// HELPER FUNCTIONS
+// SIGNATURE HEADER
 // ================================================================================================
 
-/// Returns a polynomial in Z_p[x]/(phi) representing the hash of the provided message and
-/// nonce.
-fn hash_to_point(message: Word, nonce: &NonceElements) -> Polynomial {
-    let mut state = [ZERO; Rpo256::STATE_WIDTH];
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SignatureHeader(u8);
 
-    // absorb the nonce into the state
-    for (&n, s) in nonce.iter().zip(state[Rpo256::RATE_RANGE].iter_mut()) {
-        *s = n;
+impl Default for SignatureHeader {
+    /// TODO: add docs
+    fn default() -> Self {
+        Self(0b0011_0000 + LOG_N)
     }
-    Rpo256::apply_permutation(&mut state);
+}
```

Comment:
I am not sure we should use the standardized Falcon signature header since the signature we are produce is non-standard. It is probably better to change it to something that can be used to identify our variant of the signature. I'm thinking the following could work:

- For the lower 4 bits we still use `LOG_N`.
- The format of the upper 4 bits would be `1xxx` where `xxx` would be a value identifying the hash function used for hash-to-point. For now, we'd have 2 options for this:
  - `000` for SHAKE256.
  - `001` for RPO256.

So, concretely, the default header for `RpoFalcon512` would be: `10011001`.

Reply:
@Al-Kindi-0 - I kept the header byte the way you had it (i.e., `10111001`), but we might need to figure out how to change it again if/when we migrate to `RpxFalcon512` :)

Reply:
We still have 1 free bit to use, so I think we can accommodate RPX in the current way by flipping that bit. But we can look into it further if/when we migrate :)